test/teiserver/game/balancer_server_test.exs

```diff
@@ -36,7 +37,49 @@ defmodule Teiserver.Game.BalancerServerTest do
     assert result == 4
   end
 
+  test "load some players in and run a balance pass or two" do
+    team_count = 2
+    players = create_fake_players(4)
+
+    {:ok, start_link_results} = BalancerServer.start_link(data: %{lobby_id: 1})
+
+    starting_state = GenServer.call(start_link_results, :report_state, 5000)
+    dbg(starting_state)
+
+    dbg(players)
+
+    first_balance_pass =
+      GenServer.call(
+        start_link_results,
+        {:make_balance, team_count, [], players},
+        5000
+      )
+
+    dbg(first_balance_pass)
+
+    second_balance_pass_hash_reuse =
+      GenServer.call(
+        start_link_results,
+        {:make_balance, team_count, [], create_fake_players(4)},
+        5000
+      )
+
+    dbg(second_balance_pass_hash_reuse)
+
+    assert second_balance_pass_hash_reuse.hash == first_balance_pass.hash
+
+    third_balance_pass_with_more_players =
+      GenServer.call(
+        start_link_results,
+        {:make_balance, team_count, [], create_fake_players(160)},
```

Comment:
Currently failing because player 5 does not have a valid rank.

 I'll work on the test handler to create users as needed next, but refactoring the make_balance handler to be public and allow players to be passed in really made iterating through the test process easier.

Reply:
Sounds good.
Also, having test fixtures to create users in specific state is going to be helpful in general. This is an area I'm not too familiar with.

The general plan for balance is to reduce it to a pure function that takes a list of players and their attributes like OS, playtime and ranks, along with a specification of balancing, like the number of team and the number of player per team. This may be done through a network call to another service, or a pure elixir fallback. But I want to decouple all state concerns from the balancing algorithm itself. In that sense, focusing on a public `make_balance` is probably a good move.